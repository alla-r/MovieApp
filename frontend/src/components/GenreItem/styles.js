import styled from 'styled-components';

// eslint-disable-next-line import/prefer-default-export
export const GenreItemWrapper = styled.button`
  margin-right: 10px;
  margin-bottom: 10px;
  padding: 8px 22px;
  background-color: transparent;
  color: ${(props) => props.theme.colors.secondary};
  border: ${(props) => `1px solid ${props.theme.colors.secondary}`};
  border-radius: 50px;
  font-weight: 300;
  font-size: 14px;
  line-height: 17px;
  cursor: pointer;

  &:hover {
    background-color: ${(props) => props.theme.colors.secondary};
    color: ${(props) => props.theme.colors.light};
  }
`;
