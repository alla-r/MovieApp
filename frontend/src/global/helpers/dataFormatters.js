/* eslint-disable camelcase */
const getFormattedPersonItem = (item) => {
  const formattedItem = {
    id: item.id,
    job: item.job || '',
    department: item.department || item.known_for_department,
    name: item.name,
    character: item.character,
    poster: `https://image.tmdb.org/t/p/original${item.profile_path}`,
  };

  return formattedItem;
};

const getFormattedRecommendationItem = (item) => {
  const formattedItem = {
    id: item.id,
    type: item.media_type,
    title: item.title || item.name,
    overview: item.overview,
    voteAvg: item.vote_average,
    date: item.release_date || item.first_air_date,
    poster: `https://image.tmdb.org/t/p/original${item.profile_path}`,
    backdrop: `https://image.tmdb.org/t/p/original${item.backdrop_path}`,
  };

  return formattedItem;
};

const getFormattedMediaDetails = (type, data) => {
  const formattedData = {};

  if (type === 'movie') {
    formattedData.details = {
      type,
      id: data.id,
      title: data.title,
      originalTitle: data.original_title,
      genres: data.genres,
      overview: data.overview,
      voteAvg: data.vote_average,
      status: data.status,
      poster: `https://image.tmdb.org/t/p/original${data.poster_path}`,
      backdrop: `https://image.tmdb.org/t/p/original${data.backdrop_path}`,
      productionCompanies: data.production_companies.map((item) => item.name),
      productionCountries: data.production_countries.map((item) => item.name),
      duration: data.runtime,
      revenue: data.revenue,
      budget: data.budget,
      date: data.release_date,
    };
    formattedData.cast = data.credits.cast.map((item) => getFormattedPersonItem(item));
    formattedData.crew = data.credits.crew.map((item) => getFormattedPersonItem(item));
    formattedData.recommendations = data.recommendations.results.map((item) =>
      getFormattedRecommendationItem(item),
    );
  }

  if (type === 'tv') {
    formattedData.details = {
      type,
      id: data.id,
      title: data.name,
      originalTitle: data.original_name,
      genres: data.genres,
      overview: data.overview,
      voteAvg: data.vote_average,
      status: data.status,
      poster: `https://image.tmdb.org/t/p/original${data.poster_path}`,
      backdrop: `https://image.tmdb.org/t/p/original${data.backdrop_path}`,
      productionCompanies: data.production_companies.map((item) => item.name),
      productionCountries: data.production_countries.map((item) => item.name),
      createdBy: data.created_by.map((creator) => ({
        id: creator.id,
        name: creator.name,
        job: 'Created By',
      })),
      numberOfSeasons: data.number_of_seasons,
      duration: data.episode_run_time[0],
      firstAirDate: data.first_air_date,
      lastAirDate: data.last_air_date,
    };
    formattedData.cast = data.credits.cast.map((item) => getFormattedPersonItem(item));
    formattedData.crew = data.credits.crew.map((item) => getFormattedPersonItem(item));
    formattedData.recommendations = data.recommendations.results.map((item) =>
      getFormattedRecommendationItem(item),
    );
  }

  return formattedData;
};

const getAge = (birthday, deathday) => {
  let age;
  let ageDifMs;
  let ageDate;

  if (deathday) {
    ageDifMs = new Date(deathday).getTime() - new Date(birthday).getTime();
    ageDate = new Date(ageDifMs);

    age = Math.abs(ageDate.getUTCFullYear() - 1970);
  } else {
    ageDifMs = new Date() - new Date(birthday).getTime();
    ageDate = new Date(ageDifMs);

    age = Math.abs(ageDate.getUTCFullYear() - 1970);
  }

  return age;
};

const getFormattedPersonDetails = (data = {}) => {
  const formattedDetails = {
    id: data.id,
    name: data.name,
    knownFor: data.known_for_department,
    birthday: `${data.birthday} (${getAge(data.birthday, data.deathday)} years old)`,
    deathday: data.deathday,
    placeOfBirth: data.place_of_birth,
    poster: `https://image.tmdb.org/t/p/original${data.profile_path}`,
    biography: data.biography,
    age: getAge(data.birthday, data.deathday),
  };

  return formattedDetails;
};

const getFormattedGenreList = (genreList, selectedGenresArr) => {
  const formattedGenreList = genreList.map(({ id, name }) => ({
    id,
    name,
    isChosen: selectedGenresArr.includes(id.toString()),
  }));

  return formattedGenreList;
};

const getFormattedItem = (item, type) => {
  const formattedItem = {
    id: item.id,
    type: item.media_type || type,
    date: item.release_date || item.first_air_date || 'unknown',
    title: item.title || item.name,
    description: item.overview,
    voteAvg: item.vote_average || 0,
    poster: `https://image.tmdb.org/t/p/original${item.poster_path}`,
  };

  return formattedItem;
};

const getFormattedListData = ({ page, total_pages, results }, mediaType) => {
  const formattedResultsList =
    results.length > 0 ? results.map((item) => getFormattedItem(item, mediaType)) : results;

  const formattedData = {
    page,
    totalPages: total_pages,
    results: formattedResultsList,
  };

  return formattedData;
};

const getFormattedSearchItem = (item, type) => {
  const formattedItem = {
    id: item.id,
    type: item.media_type || type,
    date: item.release_date || item.first_air_date || 'unknown',
    title: item.title || item.name,
    overview: item.overview,
    poster: `https://image.tmdb.org/t/p/original${item.poster_path}`,
  };

  return formattedItem;
};

const getFormattedSearchData = ({ page, total_pages, total_results, results }, mediaType) => {
  let formattedResultsList;

  if (mediaType === 'people' || mediaType === 'person') {
    formattedResultsList = results.map(getFormattedPersonItem);
  } else {
    formattedResultsList = results.map((item) => getFormattedSearchItem(item, mediaType));
  }

  const formattedData = {
    currentPage: page,
    totalPages: total_pages,
    results: formattedResultsList,
    totalResults: total_results,
  };

  return formattedData;
};

const getFormattedSocialMedia = ({ instagram_id, facebook_id, twitter_id }) => ({
  instagram: instagram_id ? `https://instagram.com/${instagram_id}` : null,
  facebook: facebook_id ? `https://www.facebook.com/${facebook_id}` : null,
  twitter: twitter_id ? `https://twitter.com/${twitter_id}` : null,
});

export {
  getFormattedMediaDetails,
  getFormattedGenreList,
  getFormattedListData,
  getFormattedSearchData,
  getFormattedPersonDetails,
  getFormattedSocialMedia,
};
