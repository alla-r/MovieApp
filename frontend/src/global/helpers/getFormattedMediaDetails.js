const getFormattedCreditsItem = (item) => {
  const formattedItem = {
    id: item.id,
    job: item.job || '',
    department: item.department || item.known_for_department,
    name: item.name,
    character: item.character,
    poster: `https://image.tmdb.org/t/p/original${item.profile_path}`,
  };

  return formattedItem;
};

const getFormattedRecommendationItem = (item) => {
  const formattedItem = {
    id: item.id,
    type: item.media_type,
    title: item.title || item.name,
    overview: item.overview,
    voteAvg: item.vote_average,
    date: item.release_date || item.first_air_date,
    poster: `https://image.tmdb.org/t/p/original${item.profile_path}`,
    backdrop: `https://image.tmdb.org/t/p/original${item.backdrop_path}`,
  };

  return formattedItem;
};

const getFormattedMediaDetails = (type, data) => {
  const formattedData = {};
  // console.log(type);
  // console.log(data);

  if (type === 'movie') {
    formattedData.details = {
      type,
      id: data.id,
      title: data.title,
      originalTitle: data.original_title,
      genres: data.genres,
      overview: data.overview,
      voteAvg: data.vote_average,
      status: data.status,
      poster: `https://image.tmdb.org/t/p/original${data.poster_path}`,
      backdrop: `https://image.tmdb.org/t/p/original${data.backdrop_path}`,
      productionCompanies: data.production_companies.map((item) => item.name),
      productionCountries: data.production_countries.map((item) => item.name),
      duration: data.runtime,
      revenue: data.revenue,
      budget: data.budget,
      date: data.release_date,
    };
    formattedData.cast = data.credits.cast.map((item) => getFormattedCreditsItem(item));
    formattedData.crew = data.credits.crew.map((item) => getFormattedCreditsItem(item));
    formattedData.recommendations = data.recommendations.results.map((item) =>
      getFormattedRecommendationItem(item),
    );
  }

  if (type === 'tv') {
    formattedData.details = {
      type,
      id: data.id,
      title: data.name,
      originalTitle: data.original_name,
      genres: data.genres,
      overview: data.overview,
      voteAvg: data.vote_average,
      status: data.status,
      poster: `https://image.tmdb.org/t/p/original${data.poster_path}`,
      backdrop: `https://image.tmdb.org/t/p/original${data.backdrop_path}`,
      productionCompanies: data.production_companies.map((item) => item.name),
      productionCountries: data.production_countries.map((item) => item.name),
      createdBy: data.created_by.map((creator) => ({
        id: creator.id,
        name: creator.name,
        job: 'Created By',
      })),
      numberOfSeasons: data.number_of_seasons,
      duration: data.episode_run_time[0],
      firstAirDate: data.first_air_date,
      lastAirDate: data.last_air_date,
    };
    formattedData.cast = data.credits.cast.map((item) => getFormattedCreditsItem(item));
    formattedData.crew = data.credits.crew.map((item) => getFormattedCreditsItem(item));
    formattedData.recommendations = data.recommendations.results.map((item) =>
      getFormattedRecommendationItem(item),
    );
  }

  return formattedData;
};

export default getFormattedMediaDetails;
