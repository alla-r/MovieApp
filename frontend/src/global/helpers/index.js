import {
  getFormattedMediaDetails,
  getFormattedGenreList,
  getFormattedListData,
} from './dataFormatters';
import ScrollToTop from './ScrollToTop';

export { getFormattedMediaDetails, ScrollToTop, getFormattedGenreList, getFormattedListData };

export const getFormattedItem = (item) => {
  const formattedItem = {
    id: item.id,
    type: item.media_type,
    date: item.release_date || item.first_air_date,
    title: item.title || item.name,
    description: item.overview,
    voteAvg: item.vote_average,
    poster: `https://image.tmdb.org/t/p/original${item.poster_path}`,
  };

  return formattedItem;
};

const BAR_COLORS = [
  {
    minValue: 70,
    maxValue: 100,
    path: 'rgba(141, 255, 71, 1)',
    trail: 'rgba(141, 255, 71, 0.3)',
  },
  {
    minValue: 30,
    maxValue: 69,
    path: 'rgba(255, 203, 71, 1)',
    trail: 'rgba(255, 203, 71, 0.3)',
  },
  {
    minValue: 0,
    maxValue: 29,
    path: 'rgba(255, 71, 71, 1)',
    trail: 'rgba(255, 71, 71, 0.3)',
  },
];

export const getBarColor = (type, percentageValue) => {
  const colors = BAR_COLORS.find(
    (element) => percentageValue >= element.minValue && percentageValue <= element.maxValue,
  );

  return (colors && colors[type]) || '#5B5B5B';
};

export const getFormattedDate = (dateString) => {
  const dateArray = dateString.split('-');
  const year = dateArray[0];
  const month = dateArray[1];
  const day = dateArray[2];
  const months = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ];

  return dateString ? `${months[month - 1].slice(0, 3)} ${day}, ${year}` : "";
};

export const getPercentageValue = (value, maxValue) => (value * 100) / maxValue;

export const getCrewListWithUniqueItems = (list) => {
  const names = new Set(list.map(({ name }) => name));
  const uniqueList = [];

  // eslint-disable-next-line no-restricted-syntax
  for (const uniqueName of names) {
    const jobsArray = list.filter(({ name }) => name === uniqueName).map(({ job }) => job);
    const { id } = list.find(({ name }) => name === uniqueName);
    uniqueList.push({
      id,
      job: jobsArray.join(', '),
      name: uniqueName,
    });
  }

  return uniqueList;
};

export const getFormattedYears = ({ type, date, firstAirDate, lastAirDate }) => {
  let formattedDate;

  if (type === 'movie') {
    formattedDate = date ? date.slice(0, 4) : null;
  }

  if (type === 'tv') {
    const startYear = firstAirDate ? firstAirDate.slice(0, 4) : null;

    const endYear = lastAirDate ? lastAirDate.slice(0, 4) : '...';
    formattedDate = startYear ? `${startYear} - ${endYear}` : "";
  }

  return formattedDate;
};

export const getDuration = ({ duration }) => {
  const hours = Math.trunc(duration / 60);
  const minutes = duration % 60;

  let formattedDuration = '';

  if (hours) {
    formattedDuration += `${hours}h `;
  }

  if (minutes) {
    formattedDuration += `${minutes}m`;
  }

  return formattedDuration;
};
