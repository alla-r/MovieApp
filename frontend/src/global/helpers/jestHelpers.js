/* eslint-disable react/react-in-jsx-scope */
/* eslint-disable react/jsx-filename-extension */
/* eslint-disable import/prefer-default-export */
import { ThemeProvider } from 'styled-components';
import theme from '../../theme';

export const wrapperWithTheme = (fn, tree) => fn(
  <ThemeProvider theme={theme}>
    {tree}
  </ThemeProvider>
)