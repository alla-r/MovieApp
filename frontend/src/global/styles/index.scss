*,
body,
html,
head {
  font-family: 'Montserrat', sans-serif !important;
  margin: 0px;
  padding: 0px;
  box-sizing: border-box;
}

.page-container {
  position: relative;
  min-height: 100vh;

  .content-wrap {
    padding-bottom: 160px;

    .container {
      max-width: 1480px;
      width: 90%;
      .pagination-btn {
        margin: 0 auto;
        display: block;
      }
    }
  }

  footer {
    position: absolute;
    bottom: 0;
  }
}
