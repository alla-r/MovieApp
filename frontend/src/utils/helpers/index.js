import {
  getFormattedMediaDetails,
  getFormattedGenreList,
  getFormattedListData,
  getFormattedSearchData,
  getFormattedPersonDetails,
  getFormattedSocialMedia,
  getFormattedRecommendationItem,
} from './dataFormatters';

import {
  getFormattedItem,
  getBarColor,
  getFormattedDate,
  getPercentageValue,
  getCrewListWithUniqueItems,
  getFormattedYears,
  getDuration,
  getUniqueValues,
  showNotification,
  getTokenDuration,
} from './general';

import ScrollToTop from './ScrollToTop';

export {
  getFormattedMediaDetails,
  getFormattedGenreList,
  getFormattedListData,
  getFormattedSearchData,
  getFormattedPersonDetails,
  getFormattedSocialMedia,
  getFormattedRecommendationItem,
  getFormattedItem,
  getBarColor,
  getFormattedDate,
  getPercentageValue,
  getCrewListWithUniqueItems,
  getFormattedYears,
  getDuration,
  getUniqueValues,
  showNotification,
  getTokenDuration,
  ScrollToTop,
};
