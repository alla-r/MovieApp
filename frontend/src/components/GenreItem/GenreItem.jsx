import React from 'react';
import PropTypes from 'prop-types';
import { GenreItemWrapper } from './styles';

const GenreItem = ({ onClickHandler, genre, isActive }) => (
  <GenreItemWrapper onClick={onClickHandler} className={isActive ? "active" : ""}>{genre}</GenreItemWrapper>
);

GenreItem.defaultProps = {
  isActive: false,
};

GenreItem.propTypes = {
  onClickHandler: PropTypes.func.isRequired,
  genre: PropTypes.string.isRequired,
  isActive: PropTypes.bool
};

export default GenreItem;
