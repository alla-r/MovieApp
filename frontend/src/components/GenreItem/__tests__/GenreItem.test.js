import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import GenreItem from '../GenreItem';
import { wrapperWithTheme } from '../../../global/helpers';

describe('<GenreItem />', () => {
  let defaultProps;

  beforeEach(() => {
    defaultProps = {
      onClickHandler: jest.fn(),
      isActive: false,
      genre: "Test"
    }
  });

  const component = (props) => wrapperWithTheme(shallow, <GenreItem {...props} />);

  it('should be defined', () => {
    const wrapper = component(defaultProps);

    expect(wrapper).toBeDefined();
  });

  it('should match snapshot', () => {
    const wrapper = component(defaultProps);
    const tree = renderer.create(wrapper).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should have active class', () => {
    defaultProps.isActive = true;
    const wrapper = shallow(<GenreItem {...defaultProps} />);

    expect(wrapper.find('.active')).toBeTruthy();
    expect(wrapper.find('.active')).not.toBeFalsy();
  });
});

