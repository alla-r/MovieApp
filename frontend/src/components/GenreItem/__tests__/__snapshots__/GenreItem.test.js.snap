// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GenreItem /> should match snapshot 1`] = `
<button
  className="sc-bdvvtL bDMMJq"
  onClick={[Function]}
>
  Test
</button>
`;
